map/client/client.go

```diff
@@ -0,0 +1,335 @@
+	}
+	if resp.Status == common.StatusNotFound {
+		return nil, ErrStatusNotFound
+	}
+	return resp.Payload, nil
+}
+
+// RWClient has both Get and Put
+type RWClient interface {
+	Put(addr []byte, payload []byte) error
+	Get(addr []byte) ([]byte, error)
+	PayloadSize() int
+}
+
+// ROClient only has Get
+type ROClient interface {
+	Get(addr []byte) ([]byte, error)
+	PayloadSize() int
+}
+
+// WOClient only has Put
+type WOClient interface {
+	Put(addr []byte, payload []byte) error
+	PayloadSize() int
+}
+
+// rwMap implements ReadWriteTranport using a ReadWriteCap and reference to map client
+// The idea is that several streams can share map/client and client/session but
+// have different capabilities for the underlying Streams transported by map/client
+type rwMap struct {
+	c     *Client
+	rwCap common.ReadWriteCap
+}
+
+// Get implements RWClient.Get
+func (r *rwMap) Get(addr []byte) ([]byte, error) {
+	i := r.rwCap.Addr(addr)
+	k := r.rwCap.Read(addr)
+	return r.c.Get(i, k.Sign(i.Bytes()))
+}
+
+// Put implements RWClient.Put
+func (r *rwMap) Put(addr []byte, payload []byte) error {
+	i := r.rwCap.Addr(addr)
+	k := r.rwCap.Read(addr)
+	return r.c.Put(i, k.Sign(payload), payload)
+}
+
+// PayloadSize implements RWClient.PayloadSize
+func (r *rwMap) PayloadSize() int {
+	return r.c.PayloadSize()
+}
+
+// roMap implements ReadOnlyTranport using a ReadOnlyCap and reference to map client
+type roMap struct {
+	c     *Client
+	roCap common.ReadOnlyCap
+}
+
+// Get implements ROClient.Get
+func (r *roMap) Get(addr []byte) ([]byte, error) {
+	i := r.roCap.Addr(addr)
+	k := r.roCap.Read(addr)
+	return r.c.Get(i, k.Sign(i.Bytes()))
+}
+
+// PayloadSize implements ROClient.PayloadSize
+func (r *roMap) PayloadSize() int {
+	return r.c.PayloadSize()
+}
+
+// woMap implements WOClient
+type woMap struct {
+	c     *Client
+	woCap common.WriteOnlyCap
+}
+
+// Put implements WOClient.Put
+func (w *woMap) Put(addr []byte, payload []byte) error {
+	i := w.woCap.Addr(addr)
+	k := w.woCap.Write(addr)
+	return w.c.Put(i, k.Sign(payload), payload)
+}
+
+// PayloadSize implements WOClient.PayloadSize
+func (r *woMap) PayloadSize() int {
+	return r.c.PayloadSize()
+}
+
+// ReadWrite returns a Transport using map that can read or write with Get() and Put()
+func ReadWrite(c *Client, rwCap common.ReadWriteCap) RWClient {
+	m := new(rwMap)
+	m.c = c
+	m.rwCap = rwCap
+	return m
+}
+
+// ReadOnly returns a Transport using map that can read with Get() only
+func ReadOnly(c *Client, roCap common.ReadOnlyCap) ROClient {
+	m := new(roMap)
+	m.c = c
+	m.roCap = roCap
+	return m
+}
+
+// WriteOnly returns a Transport using map that can write with Put() only
+func WriteOnly(c *Client, woCap common.WriteOnlyCap) WOClient {
+	m := new(woMap)
+	m.c = c
+	m.woCap = woCap
+	return m
+}
+
+// create a duplex using a shared secret
+func DuplexFromSeed(c *Client, initiator bool, secret []byte) RWClient {
+	salt := []byte("duplex initialized from seed is not for multi-party use")
```

Comment:
here's another example of a method that needs to be rewritten so that we can test the math separately from the io

Reply:
Where is the test for this function?

Reply:
`DuplexFromSeed` needs a unit test.

Reply:
ideally we would separate it away the pure math from functions that take a client reference as an argument.